### ISSUE TITLE
Issue with some dates not parsing properly

### ISSUE THREAD

**Issue by JWFB** [kanedata/ixbrl-parse#39]
Hi,

Had an issue with dates not parsing properly on some accounts, the accounts in questions had dates in formats such as '31/03/2023'. The error would look something like "time data '31/12/2022' does not match format '%d.%m.%y'".

All the error seemed to revolve around the format "ixt2:datedaymonthyear".  An example of accounts that throw errors would be: THOS.S.PENNY,LIMITED Company number 00093876

I did a temporary fix in the code under the 'formats.py' file and amended the 'ixtDateFormat' Class's function 'parse_value' to be the below:

    def parse_value(self, value: Union[str, int, float]) -> Optional[datetime.date]:
        if isinstance(value, str):
            value = value.lower()
            # remove ordinal suffixes with regex
            value = DATE_ORDINAL_SUFFIX_REGEX.sub(r"\1", value)
            date_formats = self._get_date_formats()
            error: Optional[Exception] = None
            for date_format in date_formats:
                try:
                    return datetime.datetime.strptime(value, date_format).astimezone().date()
                except ValueError as e:
                    error = e
                    continue
            # if we get here, we couldn't parse the date. Raise the last error
            if error:  # pragma: no cover
                try:
                    return datetime.datetime.strptime(value, "%d/%m/%Y").astimezone().date()
                except ValueError as e:
                    error = e
                    print(value)
                raise error
        msg = f"Could not parse value {value} as a date"
        warnings.warn(msg, stacklevel=2)
        return None
        
Basically just added an additional check if all other checks fail to manually try parse the date value as "%d/%m/%Y" if other formats fail before raising an error.

**Comment by JWFB** [comment]
ignore the above code, decided to just add it in here instead (still in formats.py file) to simplify things:

class ixtDateDayMonthYear(ixtDateFormat):  # noqa: N801
    format_names = (
        "datedaymonthyear",
        "ixt:datedaymonthyear",
    )
    date_format = ("%d.%m.%Y", "%d.%m.%y", "%d/%m/%Y", "%d/%m/%y")

**Comment by drkane** [comment]
Thanks - good spot.

Looking at the spec for "ixt2:datedaymonthyear" I think it just specfies a "non numeric separator", rather than just a dot format.

I think the best solution might be to be a bit more permissive in what we accept. I could add a regex to `parse_value` which will replace any non-alphanumeric characters with "-" - then we can change the date formats to all expect "-" as a separator. That should fix this issue and any other date formats with different separators that come up.

I've got a quick fix which should be deployable soon.

**Comment by drkane** [comment]
@JWFB - I've released v0.8.2 which should fix this. Let me know if it doesn't work...

**Comment by JWFB** [comment]
@drkane Thanks for coming back to me so quickly! - I'm doing quite a big bulk process at the moment, the other common one was below: 

format="ixt2:datedaymonthyearen"
Allowable formats are ('%d-%B-%Y', '%d-%B-%y') however company reg 13550020 and others contain the date format '%d-%b-%Y' i.e. '29-aug-2022' or value on raw accounts '29 aug 2022'


I don't know whether I should flag this as a separate issue, or an issue at all (because the accounts themselves are wrong) but I also get the below:

line 142, in parse_value
    return datetime.datetime.strptime(value, date_format).astimezone().date()
OSError: [Errno 22] Invalid argument

The accounts in question usually have a typo in the date, when we attempt to parse the date the above error gets thrown, which  stops us from extracting the other correct data (If I remove astimezone() from the code it works again). Example accounts would be company reg: 6954420 where the you can see "Approved by the board on 10 July 1023", there's quite a few instances of typos on company accounts which trigger this error

**Comment by drkane** [comment]
no problem @JWFB - these are really helpful issues to nail down. 

-----

For the issue with `ixt2:datedaymonthyearen` dates - I think that's a case where it's probably useful to add alternative formats, even though they're not strictly meeting the format. I've done that in version 0.9.1.

What I've also done which might be more helpful generally for dates is to allow plugins to override the existing date formats - previously a plugin trying to use the same format class would trigger an error, but now it will override instead. There's an [example in the docs of how you could add additional date formats](https://ixbrl-parse.readthedocs.io/en/latest/plugins/#override-an-existing-format).

-----

For the 1023 date I've removed `astimezone()` as it was only there to make the linter happy. Obviously this will still give the year as 1023 but I can't do much about that. This might not catch all possible parsing errors though. [It should be possible to run the parse with `raise_on_error=False`](https://ixbrl-parse.readthedocs.io/en/latest/python-module/#check-for-any-parsing-errors) - this might help collect any other errors that occur. Let me know if there's any errors that slip through this net.
